engine/grid.py

```diff
@@ -202,178 +223,435 @@ def get_neighbor_node(self, row, col, row_max, col_max):
+
+        For this traversal, the robot should traverse row by row until there are
+        no more active nodes that are encaptured by the border nodes.
+
+        When the robot reaches a the first node in a row of active nodes, it
+        will traverse down the row and turn to the next row BEFORE the edge node,
+        the last node in the row (so the turn is contained within the bounds of
+        the selected border).
+
+        Ex:  3 <- 4 <- 5
+                    ^
+             0 -> 1 -> 2
+        """
+        # The robot is traversing in the left direction, from right to left
+        # The next position in the left direction
+        new_pos = self.get_next_traversal_pos()
+
+        # The next leftmost position in row above
+        edge_column = self.edge_column_of_next_row(self.curr_pos)
+
+        if edge_column == None:  # Problem with initialization
+            self.waypoints_is_finished = True
+            return
+
+        # Column where we want to begin turning, one node closer to the inside of our shape
+        turning_column = self.get_turning_column(edge_column)
+        if is_vertical:
+            next_row = self.curr_pos[0] + 1
+        else:
+            next_row = self.curr_pos[1] + 1
+
+        # Traverse the row until it is time to turn
+        while not self.is_before_turning_column(turning_column):
+            self.waypoints.append(new_pos)
+            self.curr_pos = new_pos
+            new_pos = self.get_next_traversal_pos()
+
+        # Check whether turning to the next row is valid
+        if (is_vertical and not self.nodes[next_row][turning_column].is_active)\
+                or (not is_vertical and not (next_row < self.num_cols and self.nodes[turning_column][next_row].is_active)):
+            # Next row is out of bounds of the grid OR the node in the row/column above,
+            # where we are trying to turn to, isn't active for traversal
+            self.waypoints_is_finished = True
+
+        else:
+            # Turn to next row
+            # Create waypoints needed for a smooth turning trajectory to
+            # "guide" our robot to the next original waypoint
+            if is_vertical:
+                circle_plt = self.plot_circle((self.curr_pos[0], turning_column), (next_row, turning_column),
+                                              (self.curr_pos[0] + .5, turning_column), self.get_turn_orientation())
+            else:
+                circle_plt = self.plot_circle((turning_column, self.curr_pos[1]), (turning_column, next_row),
+                                              (turning_column, self.curr_pos[1] + .5), self.get_turn_orientation())
+            self.waypoints += circle_plt
+            # Update traversal details
+            self.switch_directions()
+            if is_vertical:
+                self.curr_pos = (next_row, turning_column)
+            else:
+                self.curr_pos = (turning_column, next_row)
```

Comment:
assuming curr_pos is (row, col), might be mislead to leave var names as turning_column for row arg and next_row for col arg